### PR TITLE
Fix annotations and spans leaking between threads

### PR DESCRIPTION
**Pull request by j-bennet**
Corresponds to this change in dask:

https://github.com/dask/dask/pull/10367

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Comment by fjetter** [comment]
add to allowlist

File: distributed/client.py [hunk 1]
```diff
@@ -3108,7 +3108,7 @@ def _graph_to_futures(
                 annotations["resources"] = resources
 
             # Merge global and local annotations
-            annotations = merge(dask.config.get("annotations", {}), annotations)
+            annotations = merge(dask.get_annotations({}), annotations)
```
Comment (crusaderky) [inline distributed/client.py:3111]:
```suggestion
            annotations = merge(dask.get_annotations(), annotations)
```

File: distributed/spans.py [hunk 2]
```diff
@@ -74,11 +74,11 @@ def span(*tags: str) -> Iterator[str]:
     if not tags:
         raise ValueError("Must specify at least one span tag")
 
-    prev_tags = dask.config.get("annotations.span.name", ())
+    prev_tags = dask.get_annotation("span.name", default_value=())
     # You must specify the full history of IDs, not just the parent, because
     # otherwise you would not be able to uniquely identify grandparents when
     # they have no tasks of their own.
-    prev_ids = dask.config.get("annotations.span.ids", ())
+    prev_ids = dask.get_annotation("span.ids", default_value=())
```
Comment (crusaderky) [inline distributed/spans.py:81]:
```python
annotation = dask.get_annotations().get("span")
prev_tags = annotation["name"] if annotation else ()
prev_ids = annotation["ids"] if annotation else ()

**Review by crusaderky — CHANGES_REQUESTED** [review]

File: distributed/spans.py [hunk 3]
```diff
@@ -74,11 +74,12 @@ def span(*tags: str) -> Iterator[str]:
     if not tags:
         raise ValueError("Must specify at least one span tag")
 
-    prev_tags = dask.config.get("annotations.span.name", ())
+    annotation = dask.get_annotations().get("span")  # type: ignore
```
Comment (crusaderky) [inline distributed/spans.py:77]:
```suggestion
    annotation = dask.get_annotations().get("span")
```
Looks like your pre-commit cache is obsolete. `rm -rf ~/.cache/pre-commit`

Comment (j-bennet) [reply distributed/spans.py]: mypy is still complaining in CI.

**Review by crusaderky — APPROVED** [review]
Ready to be merged as soon as you revert the temp changes to CI

**Comment by j-bennet** [comment]
> Ready to be merged as soon as you revert the temp changes to CI

Those changes are now reverted.

**Pull request merged**